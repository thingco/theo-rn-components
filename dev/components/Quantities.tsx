import { Text } from "./Base";
import { kmToMiles } from "../lib/quantity-functions";
import { round } from "../lib/math-functions";
import React from "react";

type DistanceUnit = "I" | "M";

type Props = {
  precision: number;
  style?: Array<import("react-native").TextStyle>;
  units: DistanceUnit;
};

type DistanceProps = {
  km: number;
} & Props;

const FormattedDistance: React.FC<DistanceProps> = ({
  km,
  units = "I",
  precision = 1,
  ...props
}: DistanceProps) => {
  const dist = units === "I" ? kmToMiles(km, precision) : round(km, precision);
  const unitSuffix = units === "I" ? "mi" : "km";

  return (
    <Text style={props.style}>
      {dist.toFixed(precision)}
      {unitSuffix}
    </Text>
  );
};

type SpeedProps = {
  kmph: number;
} & Props;

const FormattedSpeed: React.FC<SpeedProps> = ({
  kmph,
  units = "I",
  precision = 1,
  ...props
}: SpeedProps) => {
  const dist =
    units === "I" ? kmToMiles(kmph, precision) : round(kmph, precision);
  const unitSuffix = units === "I" ? "mph" : "kmph";

  return (
    <Text style={props.style}>
      {dist.toFixed(precision)}
      {unitSuffix}
    </Text>
  );
};

export { FormattedDistance, FormattedSpeed };
