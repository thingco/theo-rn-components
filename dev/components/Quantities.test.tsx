import React from "react";
import renderer from "react-test-renderer";
import { FormattedDistance, FormattedSpeed } from "./Quantities";

test("Distance renders correctly for metric", () => {
  const tree = renderer
    .create(<FormattedDistance km={42} precision={1} units="M" />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("Speed renders correctly for metric", () => {
  const tree = renderer
    .create(<FormattedSpeed kmph={84} precision={2} units="M" />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("Distance renders correctly for imperial", () => {
  const tree = renderer
    .create(<FormattedDistance km={42} precision={0} units="I" />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("Speed renders correctly for imperial", () => {
  const tree = renderer
    .create(<FormattedSpeed kmph={84} precision={1} units="I" />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
