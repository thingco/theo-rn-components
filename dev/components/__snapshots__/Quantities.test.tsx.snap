// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Distance renders correctly for imperial 1`] = `
<Text
  style={
    Array [
      Object {
        "fontFamily": "Bariol",
      },
      Object {
        "fontSize": 16,
      },
      undefined,
    ]
  }
>
  26
  mi
</Text>
`;

exports[`Distance renders correctly for metric 1`] = `
<Text
  style={
    Array [
      Object {
        "fontFamily": "Bariol",
      },
      Object {
        "fontSize": 16,
      },
      undefined,
    ]
  }
>
  42.0
  km
</Text>
`;

exports[`Speed renders correctly for imperial 1`] = `
<Text
  style={
    Array [
      Object {
        "fontFamily": "Bariol",
      },
      Object {
        "fontSize": 16,
      },
      undefined,
    ]
  }
>
  52.2
  mph
</Text>
`;

exports[`Speed renders correctly for metric 1`] = `
<Text
  style={
    Array [
      Object {
        "fontFamily": "Bariol",
      },
      Object {
        "fontSize": 16,
      },
      undefined,
    ]
  }
>
  84.00
  kmph
</Text>
`;
